PS4k/src/com/vitco/export/MagicaVoxelExporter.java

```diff
@@ -0,0 +1,123 @@
+        final int voxelsCount = data.getVisibleLayerVoxel().length;
+        final int[][] sizeMeta = this.getSizeMeta();
+
+        final BiMap<Integer, Color> palette = new BiMap<>();
+        int currentPaletteIndex = PALETTE_SIZE - 2;
+        this.setColorDefault(palette);
+
+        // Magic number
+        fileOut.writeASCIIString("VOX ");
+        // Write version
+        fileOut.writeIntRev(MV_VERSION);
+
+        //MAIN Chunk
+        fileOut.writeASCIIString("MAIN");
+        fileOut.writeIntRev(0); // Content Size
+        fileOut.writeIntRev((12 + 4 * 3) + (12 + 4 + 4 * voxelsCount) + (12 + 4 * PALETTE_SIZE));
+
+        // Size Chunk
+        fileOut.writeASCIIString("SIZE");
+        fileOut.writeIntRev(4 * 3);
+        fileOut.writeIntRev(0);
+        fileOut.writeIntRev(sizeMeta[1][0] + sizeMeta[0][0]);
+        fileOut.writeIntRev(sizeMeta[1][2] + sizeMeta[0][2]); // Maybe MV is Z-UP, Not sure but this just works
+        fileOut.writeIntRev(sizeMeta[1][1] + sizeMeta[0][1]);
+
+
+        //XYZI Chunk
+        fileOut.writeASCIIString("XYZI");
+        fileOut.writeIntRev(4 + 4 * voxelsCount);
+        fileOut.writeIntRev(0);
+        fileOut.writeIntRev(voxelsCount);
+
+        for (Voxel voxel : data.getVisibleLayerVoxel()) {
+
+            int colorId;
+            if (!palette.containsValue(voxel.getColor())) {
+                if (currentPaletteIndex < 0) {
+                    console.addLine("Error: exporter format supports only 256 colors");
+                    return false;
+                }
+                palette.put(currentPaletteIndex, voxel.getColor());
+                colorId = currentPaletteIndex + 1;
+                currentPaletteIndex -= 1;
+            } else {
+                colorId = palette.getKey(voxel.getColor()) + 1;
+            }
+
+
+            final int vx = sizeMeta[1][0] - voxel.x;
+            final int vy = sizeMeta[1][1] - voxel.y;
+            final int vz = sizeMeta[1][2] - voxel.z;
+
+            fileOut.writeByte((byte)vx);
+            fileOut.writeByte((byte)vz); // Maybe MV is Z-UP, Not sure but this just works
+            fileOut.writeByte((byte)vy);
+            fileOut.writeByte((byte)colorId);
+        }
+
+        // RGBA Chunk
+        fileOut.writeASCIIString("RGBA");
+        fileOut.writeIntRev(4 * PALETTE_SIZE);
+        fileOut.writeIntRev(0);
+
+        for (int i = 0; i < PALETTE_SIZE; i++) {
+            final Color color = palette.get(i);
+            fileOut.writeByte((byte) color.getRed());
+            fileOut.writeByte((byte) color.getGreen());
+            fileOut.writeByte((byte) color.getBlue());
+            fileOut.writeByte((byte) 255);
+        }
+        return true;
+    }
+
+    // Just want to match MV default palette "look"
+    private void setColorDefault(final BiMap<Integer, Color> palette) {
+        for (int i = 0; i < PALETTE_SIZE; i++) {
+            palette.put(i, new Color(75, 75, 75));
```

Comment:
Will change, Just didn't like to have objects as member variable when not needed (Bad habits) :D